Add Bluetooth support to Pico W

peterharperuk opened the pull request:
Add Bluetooth support to Pico W. Using BTstack with CYW43

peterharperuk committed:
rp2: Add Bluetooth support via cyw43.

Using BTstack with CYW43 for Pico W.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
extmod/btstack: Add cmake support for BTstack.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk commented on ports/rp2/CMakeLists.txt:
```diff
@@ -211,8 +213,39 @@ if(MICROPY_PY_BLUETOOTH)
         MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE=1
-        MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING=1
-        MICROPY_PY_BLUETOOTH_ENABLE_L2CAP_CHANNELS=1
+        #MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING=1
```
I think I removed this when I heard that these options were not supported?

peterharperuk replied:
I'll move inside if(MICROPY_BLUETOOTH_NIMBLE)

peterharperuk commented on ports/rp2/CMakeLists.txt:
```diff
@@ -271,6 +304,10 @@ if (MICROPY_PY_NETWORK_CYW43)
         cyw43_driver_picow
         cmsis_core
     )
+    target_include_directories(${MICROPY_TARGET} PRIVATE
```
Should be above in MICROPY_PY_BLUETOOTH_CYW43?

peterharperuk replied:
I think it's a wifi issue with the firmware inclusion. I'll leave it where it is.

/home/peterh/source/pico/micropython_latest/lib/cyw43-driver/src/cyw43_config.h:64:45: fatal error: firmware/wb43439A0_7_95_49_00_combined.h: No such file or directory

   64 | #define CYW43_CHIPSET_FIRMWARE_INCLUDE_FILE "firmware/wb43439A0_7_95_49_00_combined.h"

peterharperuk commented on ports/rp2/boards/PICO_W/mpconfigboard.cmake:
```diff
@@ -1,6 +1,11 @@
 # cmake file for Raspberry Pi Pico W
 set(MICROPY_PY_LWIP ON)
 set(MICROPY_PY_NETWORK_CYW43 ON)
+set(MICROPY_PY_BLUETOOTH_CYW43 ON)
```
Move this with the rest of the BT stuff?

peterharperuk commented on ports/rp2/mpnetworkport.c:
```diff
@@ -64,7 +64,7 @@ static void gpio_irq_handler(void) {
 
 void cyw43_irq_init(void) {
-    gpio_add_raw_irq_handler_with_order_priority(IO_IRQ_BANK0, gpio_irq_handler, CYW43_SHARED_IRQ_HANDLER_PRIORITY);
+    gpio_add_raw_irq_handler_with_order_priority(CYW43_PIN_WL_HOST_WAKE, gpio_irq_handler, CYW43_SHARED_IRQ_HANDLER_PRIORITY);
```
I think this is a bug fix from the sdk.

dpgeorge replied:
OK.  I wonder how this worked prior to this fix...

peterharperuk replied:
pass, but we had the same bug in the sdk for awhile https://github.com/raspberrypi/pico-sdk/pull/1069

dpgeorge commented on ports/rp2/main.c:
```diff
@@ -206,13 +206,16 @@ int main(int argc, char **argv) {
+        mp_bluetooth_deinit();
+        #endif
+        #if MICROPY_PY_NETWORK_CYW43 || MICROPY_PY_BLUETOOTH_CYW43
+        cyw43_deinit(&cyw43_state);
```
I'm not sure this is a good idea, to deinit the CYW43.  The idea is to retain the WiFi connection and IP address (at least) over a soft reset.

peterharperuk replied:
Ok. I can't remember why I added it. I don't think bluetooth would need it. Maybe I thought it had been forgotten. Will remove and retest.

dpgeorge commented on ports/rp2/mpbthciport.c:
```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
+void mp_bluetooth_hci_schedule_poll(void) {
     if (!events_task_is_scheduled) {
-        events_task_is_scheduled = mp_sched_schedule(MP_OBJ_FROM_PTR(&run_events_scheduled_task_obj), mp_const_none);
+        events_task_is_scheduled = mp_sched_schedule(MP_OBJ_FROM_PTR(&run_events_scheduled_task_obj), NULL);
```
The second arg here should be an object.  So at least MP_OBJ_NULL.

dpgeorge commented on ports/rp2/mpbthciport.c:
```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
+    return 0;
+}
+
+#if !MICROPY_PY_BLUETOOTH_CYW43
```
Better to not use negation and change the logic to `#if MICROPY_PY_BLUETOOTH_CYW43`.

peterharperuk replied:
- Tricky. It looks like cyw43 only needs a couple of functions in here. It doesn't need any of the functions that deal with uarts. For the time being I've wrapped this in a locally defined USES_UARTS macro

dpgeorge commented on ports/rp2/mpbthciport.c:
```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(run_events_scheduled_task_obj, run_events_scheduled_task);
 
-// Called periodically (systick) or directly (e.g. UART RX IRQ) in order to
```
Why remove this line of the comment?  Is it no longer true?

dpgeorge commented on ports/rp2/mpbthciport.c:
```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
 static int64_t mp_bluetooth_hci_timer_callback(alarm_id_t id, void *user_data) {
     poll_timer_id = 0;
-    mp_bluetooth_hci_poll_now();
+    mp_bluetooth_hci_schedule_poll();
```
I think it's better to keep the original name because it matches `stm32/mpbthciport.c`, and looks similar in name to the similar function `mp_bluetooth_hci_poll_in_ms()`.

If we wanted to rename it, maybe `mp_bluetooth_hci_poll_asap()` would be better.

peterharperuk replied:
Fair enough. I guess I made the change because the rp2 port now only supports using the scheduler (unlike the stm32 port)

dpgeorge commented on ports/rp2/mpbthciport.c:
```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
     return 0;
 }
 
+// The bt stack uses this for command timeouts, so it shouldn't be used elsewhere
```
NimBLE uses this function, so this comment seems wrong.

dpgeorge commented on ports/rp2/mpbthciport.c:
```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
 void mp_bluetooth_hci_poll_in_ms(uint32_t ms) {
+    debug_printf("mp_bluetooth_hci_poll_in_ms %lu\n", ms);
+    if (poll_timer_id > 0) {
+        cancel_alarm(poll_timer_id);
```
This looks like a bug fix.  Did you find it necessary to add this cancel?

peterharperuk replied:
I can't remember and the history from last July (!) doesn't seem very informative. It might have come from a time before interrupts were working properly. I'll try removing it.

peterharperuk replied:
Seems to work ok without it

dpgeorge commented on ports/rp2/mpbthciport.h:
```diff
@@ -30,13 +30,11 @@
-// Request new data from the uart and pass to the stack, and run pending events/callouts.
-// This is a low-level function and should not be called directly, use
-// mp_bluetooth_hci_poll_now/mp_bluetooth_hci_poll_in_ms instead.
+// Request new data and pass it to the stack, and run pending events/callouts.
```
Maybe "Request new HCI data ...".

I also think the other comment lines here should be retained.

dpgeorge commented on ports/rp2/mpbtstackport.c:
```diff
@@ -0,0 +1,155 @@
+    &btstack_run_loop_base_add_data_source,
+    &btstack_run_loop_base_remove_data_source,
+    &btstack_run_loop_base_enable_data_source_callbacks,
+    &btstack_run_loop_base_disable_data_source_callbacks,
```
Are these entries 4 needed?  They aren't on stm32.

dpgeorge commented on ports/rp2/mpbtstackport.c:
```diff
@@ -0,0 +1,155 @@
+
+void mp_bluetooth_btstack_port_deinit(void) {
+    hci_power_control(HCI_POWER_OFF);
+    // We really need hci to be closed or else we might get an bt interrupt going off and
```
"get a BT interrupt"

peterharperuk commented on ports/rp2/boards/PICO_W/mpconfigboard.h:
```diff
@@ -20,3 +20,10 @@
+
+// These aren't actually needed because we use the shared bus interface.
+#define MICROPY_HW_BLE_UART_ID -1
+#define MICROPY_HW_BLE_UART_BAUDRATE 115200
```
I can get rid of them if I remove ${BTSTACK_EXTMOD_DIR}/btstack_hci_uart.c from btstack.cmake. I guess that's ok as the hci implementation is up to the port?

dpgeorge replied:
Yes that's OK to remove the file from btstack.cmake.  That would clean things up with the use of UART here.

peterharperuk commented on ports/rp2/btstack_inc/btstack_config.h:
```diff
@@ -0,0 +1,9 @@
+#define _PICO_BTSTACK_BTSTACK_CONFIG_H
+
+#define HCI_OUTGOING_PRE_BUFFER_SIZE 4
+#define HCI_ACL_CHUNK_SIZE_ALIGNMENT 4
```
HCI_OUTGOING_PRE_BUFFER_SIZE should be there. HCI_ACL_CHUNK_SIZE_ALIGNMENT was a btstack fix found during development. I suspect you would need to update btstack. But the pico-sdk hci code will #error if it isn't defined (bit unwise in retrospect). If ACL packet is larger than Bluetooth packet buffer then it has to reduce it in size and for cyw43 this must be 4 byte aligned.

peterharperuk commented:
Note: I think I addressed the review comments so far. We're still waiting for some news on a fix for an issue from Infineon, so we'd prefer if this was not merged yet.

dpgeorge commented:
> We're still waiting for some news on a fix for an issue from Infineon, so we'd prefer if this was not merged yet.

OK.  But note that this is blocking #10635 because that PR needs to update cyw43-driver as well, which in turn requires pico-sdk 1.5.0.  Will pico-sdk need updating with the fix from Infineon?  If yes then we can wait.  If no then would be good to just update pico-sdk to 1.5.0 (and cyw43-driver) and do the minimal changes to the rp2 port to work with the new pico-sdk (that's hopefully very minimal work).  Then BT can be added as a second step, as done in this PR.

dpgeorge commented:
See #10764 which updates pico-sdk to 1.5.0 and cyw43-driver to the latest.  That's a very simple change, then this PR can be easily rebased on that once it's merged.

samveen commented:
> See #10764 which updates pico-sdk to 1.5.0 and cyw43-driver to the latest. That's a very simple change, then this PR can be easily rebased on that once it's merged.

@peterharperuk Impersonating my kid, are we there yet?

peterharperuk commented:
@samveen We're still waiting to see if we can get an update from Infineon. In the meantime, feel free to try this PR out for yourself.

FlantasticDan commented:
@peterharperuk can you provide any clarity as to what this issue is?  I've been able to get this PR working on my Pico but not quite up to feature parity with the ESP32 stack and I'm wondering if that's related...

peterharperuk commented:
It's just a stability issue. What features are missing?

FlantasticDan commented:
It seems that remote writes to a characteristic (`BLE.gattc_write`) only work if the mode option is set `1` instead of the default `0`.  It isn't the end of the world but I think it breaks part of aioble because code that worked on an ESP32 errors when writing on a Pico W.

I've also been able to get pairing and pairing with a passkey working in "Legacy Pairing" mode but enabling "LE Secure" mode so as to allow for bonding does not work, it just seems to get stuck.  I realize those options are disabled in this PR so I totally understand if that's too in the weeds for you right now.

peterharperuk commented:
@FlantasticDan Could you point me towards an example that demonstrates the problem with gattc_write mode set to zero? I believe you, but the ble_simple_central.py example seems to work okay and that appears to be using write-without-response.

FlantasticDan commented:
@peterharperuk I think I have gotten to the bottom of this and I fear I may have steered you a little off course.  My apologies.  I was seeing that my writes were erroring and assumed it was because of that write mode.  Both write modes are working as expected.  What's actually happening is the write response handler (only used when mode = `1`) never get's called because the `_IRQ_GATTC_WRITE_DONE` event is not emitting the correct `value_handle`.  My error was interpreting the timeout that I was getting as a "failed to write" and not what it actually was: "the write was never confirmed."

To test this I'm running [ble_simple_peripheral.py](https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_simple_peripheral.py) on a TinyS3 and [ble_simple_central.py](https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_simple_central.py) on a Pico W using this PR's build of micropython. I've made a [few changes to ble_simple_central.py](https://gist.github.com/FlantasticDan/8207b9f65aa0a1bfc4faab6cc40e3f16) to test this bug:
```python
# ble_simple_central.py
        elif event == _IRQ_GATTC_WRITE_DONE:
            conn_handle, value_handle, status = data
            print("TX complete", value_handle) # line 147 modified print statement to see the value_handle of the response event
# . . .
    with_response = True # line 228, was False, changed to True

    i = 0 # line 230
    print("RX Handle:", central._rx_handle) # added this statement to confirm outbound write value_handle
    while central.is_connected():
# . . .
```

Running this test sees all writes successfully reach the TinyS3 but where it's interesting is in the Pico W stdout:
```
MicroPython v1.19.1-863-g3e4244f01-dirty on 2023-02-24; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> import ble_simple_central
>>> ble_simple_central.demo()
service (64, 10, 14, UUID(0x1800))
service (64, 15, 18, UUID(0x1801))
service (64, 19, 65535, UUID('6e400001-b5a3-f393-e0a9-e50e24dcca9e'))
Connected
RX Handle: 24
TX 0_
RX <memoryview>
RX <memoryview>
RX <memoryview>
TX complete 65535
```

The writes occur at `value_handle` 24 but their responses come in as value handle 65535.  I suspect this 65535 is a hardcoded bug somewhere (haven't gotten to look very hard for it yet) because the same handle is also returned for all writes to all handles in my Blackmagic Camera use case as well.  So it seems to me the Pico W is correctly receiving the write response but is delivering it to the user code in correctly.  It's also possible it's a BTstack bug and not an issue with this PR, I don't have an alternate BTstack micropython board to test on.  I'd be curious to see if you can replicate this behavior? Or maybe I'm totally misunderstanding and this is the expected Bluetooth behavior (I'm new to all this BT stuff). Thanks!

jimmo commented:
@FlantasticDan this is a known issue with our btstack bindings. I have been working on this in the past week (in between other things).

v1.20 will include feature parity with NimBLE for all the GATT functionality (and hopefully also l2cap channels), which will mean that aioble will work identically.

jimmo commented:
@FlantasticDan the background is that btstack doesn't provide the value handle in the event, so it's up to us to track it.

It would be nice if we could make btstack do this but 
I discussed it with the btstack developer here https://github.com/bluekitchen/btstack/issues/439 and we concluded that it wasn't too hard for micropython to do it instead. (Based on what I learnt there it seems we can actually simplify our bindings a bit).

I'll have a PR for this ready soon.

@peterharperuk Yes, just to confirm this is completely unrelated to the issue you're tracking with infineon.

FlantasticDan commented:
@jimmo that's awesome, thanks for all your work on that.  I think I just came across that issue in the source code 😅
https://github.com/micropython/micropython/blob/2e4dda3c2072c4de15ac8d51ddb471b9c3833db7/extmod/btstack/modbluetooth_btstack.c#L432-L439

Sorry for steering you astray @peterharperuk

peterharperuk commented:
np

felixdoerre commented:
Hi there, I've tried out this PR, and noticed two bugs/instabilities, that could be addressed here or resolved as separate PRs/issues, as they are closely related but not strictly implied by this change:

- Randomly missing hci-callbacks: When I use bluetooth and enable bonding, I noticed that btstack sometimes misses to send replies, even though an appropriate timer, with zero delay is in `mp_btstack_runloop_timers`, the main loop is not executed. I traced this down to a bug (I believe) in `mp_bluetooth_hci_poll_in_ms`. What happens in my case: btstack creates multiple timers, one after 30s (probably to abort the bonding), and one for immediate execution, this causes `mp_bluetooth_hci_poll` to register a new alarm in roughly 30s. But before that alarm expires new hci package are handled and processed, creating a new "immediate" timer, and causing `mp_bluetooth_hci_poll` to register another 29.x-s alarm. This repeats until all pico timer slots are full with roughly 30s-timers, and adding a new "1ms"-timer fails, causing the bluetooth stack to hang until the pairing times out. I would suggest to resolve this like this:
```diff
 void mp_bluetooth_hci_poll_in_ms(uint32_t ms) {
+    if(poll_timer_id != 0){
+      cancel_alarm(poll_timer_id);
+    }
     poll_timer_id = add_alarm_in_ms(ms, mp_bluetooth_hci_timer_callback, NULL, true);
 }
```

- When I use random bluetooth addresses, the btstack fails to initialize with 25% probability. This is caused, by the corresponding code not setting the correct address type flags in the configuration used here:
```diff
@@ -592,26 +601,26 @@ STATIC void set_random_address(void) {
         bd_addr_t static_addr;
 
         #if MICROPY_BLUETOOTH_USE_MP_HAL_GET_MAC_STATIC_ADDRESS
 
         DEBUG_printf("set_random_address: Generating static address using mp_hal_get_mac\n");
         mp_hal_get_mac(MP_HAL_MAC_BDADDR, static_addr);
-        // Mark it as STATIC (not RPA or NRPA).
-        static_addr[0] |= 0xc0;
 
         #else
 
         DEBUG_printf("set_random_address: Generating random static address.\n");
         btstack_crypto_random_t sm_crypto_random_request;
         volatile bool ready = false;
         btstack_crypto_random_generate(&sm_crypto_random_request, static_addr, 6, &btstack_static_address_ready, (void *)&ready);
         while (!ready) {
             MICROPY_EVENT_POLL_HOOK
         }
 
         #endif // MICROPY_BLUETOOTH_USE_MP_HAL_GET_MAC_STATIC_ADDRESS
+        // Mark it as STATIC (not RPA or NRPA).
+        static_addr[0] |= 0xc0;
 
         DEBUG_printf("set_random_address: Address generated.\n");
         gap_random_address_set(static_addr);
     }
```

peterharperuk commented:
Thanks @felixdoerre those changes look reasonable to me.

cristi8 commented:
I am not very familiar with the repo and needed files, but should the `board.json` file  ( https://github.com/micropython/micropython/blob/master/ports/rp2/boards/PICO_W/board.json ) specify that the board supports BLE in the `features` section? (as does for ESP32). Is the file used anywhere?

Anyway, thanks for implementing this right after the pico 1.5.0 sdk came out. Can't wait to have it available in the nightly build

peterharperuk commented:
@cristi8 Looks like you are correct. Thanks.

dpgeorge commented:
@peterharperuk would be good to rebase this to keep it current.  Especially now since pico-sdk and cyw43-driver have been updated in master.  Are you able to do that?

Would also be good to incorporate the above fixes mentioned by @felixdoerre  and @cristi8

peterharperuk commented:
Yes - will do.

peterharperuk committed:
rp2/mpbthciport: Cancel existing alarms.

Cancel any existing poll alarm before add a new one.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
extmod/btstack: Fix marking of static addresses in set_random_address.

Marking address as static was not applied to all code paths.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2/boards/PICO_W: Enable Bluetooth Low Energy support.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk commented:
Had lots of issues rebasing this change. There seems to be an issue with this change...

commit c80e7c14e6305e50e3b39f97172d4d8fe1214d3b
Author: cpottle9 <cpottle9@outlook.com>
Date:   Tue Mar 7 15:24:36 2023 -0800

    rp2: Allocate GC heap from unused RAM.

assertion "ATB_GET_KIND(area, block) == AT_HEAD" failed

peterharperuk commented:
I have updated the PR. However for the time being you'll have to add the following to lib/pico-sdk https://github.com/raspberrypi/pico-sdk/pull/1309

peterharperuk commented:
Updated to rename cyw43_alloc to cyw43_malloc

dpgeorge commented on extmod/network_cyw43.c:
```diff
@@ -233,6 +241,10 @@ STATIC mp_obj_t network_cyw43_connect(size_t n_args, const mp_obj_t *pos_args, m
 
+    #ifdef hal_check_network_allowed
+    hal_check_network_allowed(MP_QSTR_WLAN);
+    #endif
```
I don't think this is needed, because the WLAN must be active before you can call `.connect()`.

peterharperuk replied:
Not exactly true. It only has to be active the first time. The following works. Maybe this is a bug that should be fixed (in the driver)?

import network
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.active(False)
wlan.connect('Pi Towers Guest', 'welcometopitowers') #<-- SUCCEEDS

dpgeorge replied:
That's a bug...

But at the same time, it'd probably be nice if `wlan.connect()` automatically activated the interface.  Well, that's out of scope for what we are doing here.

peterharperuk replied:
See other comment - I disagree, wlan.active(True) just loads the driver. For STA mode it's connnect/disconnect that matters. I've tried to clear up the behaviour of the API with the latest change.

dpgeorge commented on extmod/network_cyw43.c:
```diff
@@ -280,13 +292,19 @@ STATIC mp_obj_t network_cyw43_connect(size_t n_args, const mp_obj_t *pos_args, m
     }
+    #ifdef hal_set_network_active
+    hal_set_network_active(MP_QSTR_WLAN, true);
+    #endif
```
As above, I don't think this is needed.

peterharperuk replied:
I'm going to disagree. We need some sort of check in here for network.STA_IF. All active(True) does for STA is load the driver. And active(False) does nothing for STA. It's connect and disconnect that matters.

dpgeorge replied:
Yes you're right.  There really needs to be a check in connect/disconnect that it should fail if the interface is not activated.

But that's a bug to fix at a later stage.  For now I think what you've done is correct.

dpgeorge commented on extmod/network_cyw43.c:
```diff
@@ -280,13 +292,19 @@ STATIC mp_obj_t network_cyw43_connect(size_t n_args, const mp_obj_t *pos_args, m
     cyw43_wifi_leave(self->cyw, self->itf);
+    #ifdef hal_set_network_active
+    hal_set_network_active(MP_QSTR_WLAN, false);
+    #endif
```
As above, shouldn't be needed, because you call `.active(False)` to turn of the WiFi interface.

peterharperuk replied:
See other comments.

dpgeorge commented on extmod/network_cyw43.c:
```diff
@@ -121,8 +121,16 @@ STATIC mp_obj_t network_cyw43_active(size_t n_args, const mp_obj_t *args) {
         cyw43_wifi_set_up(self->cyw, self->itf, mp_obj_is_true(args[1]), country);
+        #ifdef hal_set_network_active
+        hal_set_network_active(MP_QSTR_WLAN, mp_obj_is_true(args[1]));
+        #endif
```
This needs to take into account both STA and AP.  Either or both can be enabled, but you only want to release access to the network (ie allow BT) if both STA and AP are turned off.

peterharperuk replied:
Yes - I've changed to use STA/AP and I'm using a bit mask so STA and AP are handled properly.

dpgeorge commented on ports/rp2/mpconfigport.h:
```diff
@@ -281,4 +281,10 @@ extern void lwip_lock_release(void);
+
+// Prevent BT+Wifi being active at the same time
+#define hal_check_network_allowed rp2_check_network_allowed
+#define hal_set_network_active rp2_set_network_active
```
I think these macros should be called something like `mp_hal_network_check_allowed()` and `mp_hal_network_set_active()`.

dpgeorge commented:
I guess this is blocked waiting on pico-sdk v1.5.1?

peterharperuk commented:
> I guess this is blocked waiting on pico-sdk v1.5.1?

Unfortunately, yes. Not entirely sure when we'll release that unfortunately.

dpgeorge commented:
There seems to be a problem checking that BLE is allowed or not.  If I activate BLE then soft reset (without deactivating), I cannot activate it again, eg:
```
MicroPython v1.19.1-998-g25347c668-dirty on 2023-03-22; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> import bluetooth
>>> bluetooth.BLE()
<BLE>
>>> _.active(1)
True
>>> 
MPY: soft reboot
MicroPython v1.19.1-998-g25347c668-dirty on 2023-03-22; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> import bluetooth
>>> bluetooth.BLE()
<BLE>
>>> bluetooth.BLE().active(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Network blocked
>>> 
```
Also, the "Network blocked" error is raised if the BLE is activated twice in a row.

peterharperuk commented:
I think I've fixed those issues. Regarding the cyw43_malloc change, it might be some weeks (maybe longer) before we release a new sdk. Would you be prepared to update to an sdk off develop or we could create a new temporary branch containing the change?

dpgeorge commented:
> Regarding the cyw43_malloc change, it might be some weeks (maybe longer) before we release a new sdk. Would you be prepared to update to an sdk off develop or we could create a new temporary branch containing the change?

Rather than updating to a develop commit of pico-sdk, I think it'd prefer to leave a little bit of RAM for the C heap, so the standard `malloc`/`free` can actually work.  That requires only a small change to our custom ld script.

Gadgetoid commented:
I'm looking into avoiding the 1k gap, but the approach isn't pretty. I have a similar issue with C++ code needing some space for heap allocations, and rewriting that to use MicroPython's heap has been relatively successful- with some C++ related caveats I have yet to overcome.

If we keep allocating chunks of RP2 RAM for various modules and extensions that the user *might not* use, we're going to end up with none left for Python code :laughing: 

Pico SDK uses `--wrap=malloc` etc to overload memory allocation functions with mutex-locked thread-safe replacements. If we override the `pico_malloc` library and replace those wrapped functions with our own, it's possible to globally redirect all `malloc`, `calloc`, `free` and (potentially) `realloc` usage to the MicroPython heap, without any special case-by-case treatment for C extensions.

I discuss this approach more here: https://github.com/micropython/micropython/issues/11116

I have created an implementation on top of your Bluetooth branch here: https://github.com/pimoroni/micropython/commit/b53d8c2d70145aa91249034d5d1873822f00576b

This does not include the multicore mutex locking (wouldn't `tracked_calloc` have to handle that, or is there a GIL anyway?)

But it does appear to work.

Running `ble_simple_peripheral.py` with debug output, gives me:

```
malloc: 340
malloc: 600
free: 536957896
free: 536958248
Starting advertising
```

Showing that, indeed, Bluetooth is allocating on MicroPython's heap successfully.

The example works, pairs to my phone and starts streaming out numbers.

Unfortunately this isn't exactly portable outside of the RP2 port - or maybe the same approach will turn out to be easier? - but then this is all rather RP2 specific anyway?

peterharperuk commented:
Great - this sounds better than my somewhat hacky approach. The 1KB reservation was just a temporary measure until we can get the sdk updated. The BT mallocs might be able to be removed in future (I don't want to change the code too much yet until the remaining stability issue is resolved). The only other malloc is in the firmware decompression code if that ever gets merged https://github.com/georgerobotics/cyw43-driver/pull/68

Gadgetoid commented:
Seems to be the case with BT that the mallocs are small and benign enough that a solution like this has minimal risk of making things heinously unstable. Glad my amateur tinkering could potentially be of some use :laughing:

peterharperuk commented:
Yes, there are two mallocs used as temporary buffers to load the BT firmware.

Gadgetoid commented:
I have updated my branch to replace my `add_library` hack with the official (though possibly undocumented) method of disabling a Pico SDK component (mentioned here: https://github.com/raspberrypi/pico-sdk/pull/745):

```patch
-add_library(pico_malloc INTERFACE)
-add_library(pico_malloc_headers INTERFACE)
+set(SKIP_PICO_MALLOC 1)
```

samveen commented:
While not exactly relevant to the change, is it possible to save the latest build artifact of the `rp2 port/build` check/test somewhere? It'll give those of us without access to a build environment the ability to test these changes, without needing to set up a full environment.

dpgeorge commented:
Regarding malloc and reserving heap memory: I want to keep this as simple as possible, but also make sure it is robust.  I think reserving 1k is not enough because there's overhead for malloc (I think 8 bytes per allocation) and also sbrk may try to increase the heap boundary more than 1k to start with.

I tried the following to allocate 4k to the C heap:
```diff
--- a/ports/rp2/memmap_mp.ld
+++ b/ports/rp2/memmap_mp.ld
@@ -243,18 +243,19 @@ SECTIONS
     } > FLASH
 
     /* stack limit is poorly named, but historically is maximum heap ptr */
-    __StackLimit = ORIGIN(RAM) + LENGTH(RAM);
+    __StackLimit = __bss_end__ + 4096;
+    __StackLimit2 = ORIGIN(RAM) + LENGTH(RAM);
     __StackOneTop = ORIGIN(SCRATCH_X) + LENGTH(SCRATCH_X);
     __StackTop = ORIGIN(SCRATCH_Y) + LENGTH(SCRATCH_Y);
     __StackOneBottom = __StackOneTop - SIZEOF(.stack1_dummy);
     __StackBottom = __StackTop - SIZEOF(.stack_dummy);
     /* Define start and end of GC heap */
-    __GcHeapStart = __bss_end__;
-    __GcHeapEnd   = __StackLimit;
+    __GcHeapStart = __StackLimit;
+    __GcHeapEnd   = __StackLimit2;
     PROVIDE(__stack = __StackTop);
 
     /* Check if data + heap + stack exceeds RAM limit */
-    ASSERT(__StackLimit >= __HeapLimit, "region RAM overflowed")
+    ASSERT(__StackLimit2 >= __HeapLimit, "region RAM overflowed")
 
     /* Check GC heap is at least 128 KB */
     /* On a RP2040 using all SRAM this should always be the case. */
```
That seems to work well, because now `_sbrk` knows that it's maximum heap extent is 4k (between `__bss_end__` and `__StackLimit__`).

peterharperuk commented:
ok - do you want to push that, or should I do it?

peterharperuk commented on ports/rp2/cyw43_configport.h:
```diff
@@ -33,7 +33,12 @@
 #include "pendsv.h"
 
+#if MICROPY_PY_BLUETOOTH_CYW43
+#define CYW43_CHIPSET_FIRMWARE_INCLUDE_FILE "wb43439A0_7_95_49_00_combined.h"
```
This duplicates the definitions in the driver? I think we can remove the definition of CYW43_CHIPSET_FIRMWARE_INCLUDE_FILE from here

dpgeorge replied:
Yes, remove it if possible.  But it may require changing the include directories to find `firmware/wb43439A0_7_95_49_00_combined.h`

dpgeorge replied:
I removed it.  The include directory was already there.

dpgeorge commented:
> ok - do you want to push that, or should I do it?

I have made the change, for the C heap.

I have also cleaned up the commits, and cleaned up a few other bits and pieces of the code.

peterharperuk commented:
Looks good. And it seems to be working okay for me

hex007 commented:
I was testing this and after approximately 30 secs connection is being dropped for Pico W. Tried the same with ESP32 and connection persists and i had to manually disconnect. I have attached the source below. One custom thing that i have changed is MTU = 512 for my usecase.

Times mentioned are in millis
[disconnectDemo.txt](https://github.com/micropython/micropython/files/11150850/disconnectDemo.txt)

dpgeorge commented:
> I was testing this and approximately after 30 secs connection is being dropped for Pico W. Tried the same with ESP32 and connection persists and i had to manually disconnect. I have attached the source below. One custom thing that i have changed is MTU = 512 for my usecase.

@hex007 I tried to reproduce your problem but could not.  I made a simple BLE peripheral which advertised the UUID 0x3143 and accepted connections, and then ran your `disconnectDemo.py` script (after renaming it from .txt) on a Pico W, and also a PYBD-SF2 for comparison.  Both behaved the same and stayed connected to the BLE peripheral for a long time, until I manually disconnected them.

I see that in the first log for Pico W it has a `_IRQ_CONNECTION_UPDATE` event.  Does this event always occur?  It's not there for the esp32 log.

I suspect that the peripheral is deliberately disconnecting the Pico W (acting as a central) for some reason.  Can you provide more information about this peripheral?  If you have 2x Pico W's, can you try implementing the peripheral yourself on the second Pico W and then get the first to connect to it?

hex007 commented:
I have Peripheral as an android app. Let me create a bare bones version of it and try reproducing it.  

I am updating the MTU to 512 on Android and this is reflected like this on Esp32

```
Event is 5
'-> pairs -> [(1, b'\x02'), (9, b'Panda'), (3, b'C1')]

Event is 6
()

Event is 7
Peripheral connected @ 4575113
Addr type 1 : addr 7154690e80b7

Event is 27
(0, 6, 0, 500, 0)

Event is 21
(0, 512)

Event is 27
(0, 40, 0, 256, 0)
```

dpgeorge commented:
@hex007 MTU exchange with BTstack (the Bluetooth stack that the Pico W uses) is not yet fully supported in MicroPython (whereas it is with esp32 and stm32 which use the NimBLE stack).  So please try for now without changing the MTU.  MTU exchange will be fixed shortly.

dpgeorge commented:
@peterharperuk I pushed two commits to fix and simplify the wifi+bt mutual exclusion checks.  There was a problem where if wifi was connected and the Pico W did a soft reset, then wifi would remain connected and you could also start up BLE (wifi remaining connected over a soft reset is a feature).

So I fixed that.  I also reverted the API changes/fixes that you made to WLAN connect/disconnect (to automatically activate/deactivate).  Those changes need to be done separately and discussed, and made consistent with all other WLAN interfaces (eg esp32).

hex007 commented:
@dpgeorge  The connection is still being dropped after removing custom MTU config on both micropython `# ble.config(mtu=512)` and Android


Additionally to be sure that my code is not at fault, I used [nRF connect app](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp) to create a service with UUID(0x3143) and these are the logs on pico w with the same code without changing ble.config
```
>>> Running disconnectDemo.py
 
 
>>> 
Event is 5
'-> pairs -> [(1, b'\x02'), (9, b'Panda'), (3, b'C1')]

Event is 6
()

Event is 7
Peripheral connected @ 7357681
Addr type 1 : addr 6e4f00bb04e6

Event is 8
Peripheral disconnected @ 7387849
Addr type 255 : addr 000000000000
Connection last 30168
```

dpgeorge commented:
> Additionally to be sure that my code is not at fault, I used [nRF connect app](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp) to create a service with UUID(0x3143) and these are the logs on pico w with the same code without changing ble.config

Can you please detail the exact steps to create the service on nRF connect, so I (or someone else) can try and reproduce the issue.

Also it looks like you changed `disconnectDemo.py` (to print addr types), so please also upload that exact script to test with.

hex007 commented:
[disconnectDemo.txt](https://github.com/micropython/micropython/files/11157448/disconnectDemo.txt)

On nRF connect
- Install nRF connect from Play store
- Go to the Advertiser tab
- Click on the Red `+`  button in the bottom right corner
- Enter display name as "test"
- Click on `Add Record` -> `Service UUID`
- Enter Service UUID as `3143` -> OK
- Enable `Connectable` checkmark
- Click on OK to create Service
- Enable Service by clicking on toggle button
- Leave all defaults as they are and click `OK`
- See that the check mark is now enabled
- Run disconnectDemo file attached on picoW microbython build (MicroPython v1.19.1-863-g3e4244f01 on 2023-04-04; Raspberry Pi Pico W with RP2040)

peterharperuk commented:
@hex007 How long did you have to wait for the disconnection? Mine seems to be running ok. I did build micropython myself.

```
Event is 5
'-> pairs -> [(1, b'\x1a'), (10, b'\x0c'), (255, b'L\x00\x10\x07\x18\x1fv\x0c\xb9`X')]

Event is 5
'-> pairs -> [(1, b'\x1a'), (10, b'\x0c'), (255, b'L\x00\x10\x06\x13\x1aJZ\xe1\x8c')]

Event is 5
'-> pairs -> [(1, b'\x1a'), (255, b'L\x00\t\x06\x03\xb5\n\x03\x0f\xf2')]

Event is 5
'-> pairs -> [(255, b'\x06\x00\x01\t \x02\xbd.F\x8c\x06\x8b\xda\x8fj\x94]\xf7\x18\xb3P\xe8~\x04*K\x9c+\xa9')]

Event is 5
'-> pairs -> [(255, b'\x06\x00\x01\t \x02\xd2\x8f\xddB\xdd\xeb7\xaa\xbbn\x1dn\xcdn_J[h\xa6\x8a3\x9cH')]

Event is 5
'-> pairs -> [(255, b'L\x00\x07\x19\x01\x14 \x01\x98\x8f\x00\x00\x08\x1aa\x88\xe0\xb7\xd4C\x9b6\xcc\xb7\xd0%4\xe2\x89')]

Event is 5
'-> pairs -> [(255, b'\x06\x00\x01\t \x02/\xff\xa8\x07\xa2\x9a \xc927\xaap"\t\'\xc3\xe0Mz\x84\xab\x87<')]

Event is 5
'-> pairs -> [(1, b'\x1a'), (255, b'L\x00\t\x06\x030\n\x03\x0fr')]

Event is 5
'-> pairs -> [(1, b'\x02'), (3, b'C1')]

Event is 6
()

Event is 7
Peripheral connected @ 44558
Addr type 1 : addr 52cb3f1136ad

```

ph1lj-6321 commented:
We are the files for building this port ?

dpgeorge commented:
Thanks @hex007 for the detailed instructions to reproduce the issue.  I can now reproduce it with an Android phone running nRF Connect as a peripheral.

In my case the phone disconnects the Pico W after 15 seconds, every time I run it.

After the two devices connect, the phone (nRF Connect) tries to discovers services on the central (Pico W).  The Pico W sees this discover request and returns a response but then something happens and the phone never sees the response.  The phone then does two more attempts to discover services before disconnecting the Pico W.

Running the same code on PYBD-SF2W (which also has a cyw43) it works just fine, the phone can discover services on the PYBD.

Note: in order to make this work with BTstack I needed to add the following line after activating the BLE:
```python
ble.gatts_register_services(())
```
This ensures there is at least a basic set of services (GAP name) for the phone to discover.

I then implemented the nRF Connect behaviour in MicroPython running on another BLE enabled board.  This is simple peripheral code that does a `ble.gattc_discover_services()` after the peripheral is connected to.  That works just fine with Pico W, it retrieves the basic set of services, including the GAP name of the Pico W.

So everything seems to work except when the Pico W connects as a central to the Android phone.  At the moment I'm not sure how to debug this any further.

dpgeorge commented:
@peterharperuk while debugging above, I found that it's possible to enable synchronous HCI mode for BTstack, by not implementing `hci_transport_cyw43_can_send_now` and instead just putting in NULL for that entry in `hci_transport_cyw43` (see `src/rp2_common/pico_cyw43_driver/btstack_hci_transport_cyw43.c` in `pico-sdk`).  Then in `hci_transport_cyw43_send_packet()` you don't need to send the `HCI_EVENT_TRANSPORT_PACKET_SENT` packet after a successful call to `cyw43_bluetooth_hci_write`.

I wonder if you considered that already (using synchronous mode)?

hex007 commented:
@dpgeorge  If I add `ble.gatts_register_services(())` after assigning irq like this `ble.irq(irq)`  connection is dropped in 15 secs else 30 secs

```
>>>  Running disconnectDemo.py
 
 
>>> 
Event is 5
'-> pairs -> [(1, b'\x06'), (6, b'\xe8\xbfl\xcf\x17\x8b2\xb8LHj\xe5\x00\xff\xba\xab'), (22, b'\n\x18$\x04\xb4I\x03')]

Event is 5
'-> pairs -> [(255, b'\x06\x00\x01\t "\x96Uu\xbf\xa5\xf4@\xbaNl\xb9\x98"\xf3@\xd4\x028V0i\xf5W')]

Event is 5
'-> pairs -> [(1, b'\x02'), (9, b'Panda'), (3, b'C1')]

Event is 6
()

Event is 7
Peripheral connected @ 791573
Addr type 1 : addr 6e53cea6cb31

Event is 27
(64, 6, 0, 500, 0)

Event is 8
Peripheral disconnected @ 806911
Addr type 255 : addr 000000000000
Connection last 15338
 Running disconnectDemo.py
 
 
>>> 
Event is 5
'-> pairs -> [(255, b'L\x00\x12\x02\x00\x02')]

Event is 5
'-> pairs -> [(1, b'\x02'), (9, b'Panda'), (3, b'C1')]

Event is 6
()

Event is 7
Peripheral connected @ 825874
Addr type 1 : addr 6e53cea6cb31

Event is 27
(64, 6, 0, 500, 0)

Event is 8
Peripheral disconnected @ 841220
Addr type 255 : addr 000000000000
Connection last 15346
 Running disconnectDemo.py
 
 
>>> 
Event is 5
'-> pairs -> [(1, b'\x02'), (9, b'Panda'), (3, b'C1')]

Event is 6
()

Event is 7
Peripheral connected @ 850752
Addr type 1 : addr 6e53cea6cb31

Event is 8
Peripheral disconnected @ 881938
Addr type 255 : addr 000000000000
Connection last 31186
```

dpgeorge commented:
I have been able to integrate NimBLE into the Pico W as the Bluetooth stack.  Using NimBLE the `disconnectDemo.py` test works, the Android phone is able to discover the basic set of characteristics of the Pico W, and does not disconnect.

The current state is:
- Pico W + BTstack: doesn't work
- Pico W + NimBLE: works
- PYBD-SF2 (using cyw43) + BTstack: works

So this looks like an issue with the BTstack integration on Pico W.  I tried to further investigate that by making the Pico W have a virtual UART for the Bluetooth HCI, and using stm32's `mpbstackport.c` (which is known to work with PYBD-SF2 as above), but that still didn't fix the problem.

hex007 commented:
@dpgeorge can you share your branch for nimble on picoW ? I would like to test it for my usecase completely. This is if it is not too much of an ask

dpgeorge commented:
> can you share your branch for nimble on picoW ?

The most recent commit on this PR allows using NimBLE.  You just have to enable it in `ports/rp2/boards/PICO_W/mpconfigboard.cmake`.  Also make sure the NimBLE submodule is checked out.

But note that it looks like NimBLE can fail as well in certain cases, so the bug here is most likely at a lower layer than the Bluetooth stack, ie inside the cyw43.

robert-hh commented:
@dpgeorge I just compiled it for the PICO_W. Reason: @iabdalkader and me are trying to find the reason for a lock-up during bluetooth_deinit(), which happens with the Arduino Nano RP2040 and the trial SAMD port as well, when running the multi_bluetooth tests. When the lock-up occurs, the code sticks at the end of the function `ble_hs_stop_done()` at the code lines
```
    SLIST_FOREACH(listener, &slist, link) {
        listener->fn(status, listener->arg);
    }
```
processing the list `ble_hs_stop_listeners`, which at that point is corrupted, in that the `next` item of one element points to itself. The LENGTHY discussion about that is at the end of https://github.com/micropython/micropython-lib/pull/555.
That does not seem to happen with the PICO_W  + nimble package, but the bug is timing sensitive, and that package may have a more lucky timing.

peterharperuk commented:
> I wonder if you considered that already (using synchronous mode)?

I hadn't. But I'm advised that it would not be correct to use synchronous mode. Evidentially synchronous mode is not used much, all ports should use async mode going forward. The cyw43 firmware has a 4KB circular buffer for HCI commands so it's not really always ready to receive new commands.

peterharperuk commented:
I can reproduce this now.

dpgeorge commented:
> The cyw43 firmware has a 4KB circular buffer for HCI commands so it's not really always ready to receive new commands.

But `hci_transport_cyw43_can_send_now()` always returns true, and `hci_transport_cyw43_send_packet()` always succeeds.

The reason I suggested synchronous mode is because `hci_transport_cyw43_send_packet()` always sends the HCI packet and immediately responds with `HCI_EVENT_TRANSPORT_PACKET_SENT`, so it is effectively synchronous.  Or am I missing something?

peterharperuk commented:
> The reason I suggested synchronous mode is because `hci_transport_cyw43_send_packet()` always sends the HCI packet and immediately responds with `HCI_EVENT_TRANSPORT_PACKET_SENT`, so it is effectively synchronous. Or am I missing something?

I'm just going on what Matthias said to me. I think the BTstack behaves differently if hci_transport_cyw43_can_send_now is not implemented?

peterharperuk commented:
Note, repro instructions are in here https://github.com/micropython/micropython/pull/10739#issuecomment-1497231635
But the script is slightly different (added call to ble.gatts_register_services() to start the acl server) 
[disconnectDemo.zip](https://github.com/micropython/micropython/files/11220576/disconnectDemo.zip)

hex007 commented:
@peterharperuk I built with the changes suggested and it doesnt disconnect for both, disconnectDemo and my usecase.  Though mtu customization causes the connection to be dropped in 30 secs

peterharperuk commented:
Rebased and fixed some build issues with latest pico-sdk and btstack

peterharperuk commented:
oh and latest cyw43-driver

ZacharyInDallas commented:
I noticed you guys are starting up on this again.  Any ballpark estimate on when it will be ready to test with?  

Will it have pairing and bonding?

If it doesn’t offer pairing and bonding, will aioble provide that?

How will I know when it is ready?

DuaneKaufman commented:
All I want to be able to do with it (for now) is listen to advertisements with my RP2040 Badger board.

dpgeorge commented:
@peterharperuk I think we can now drop the commits that prevent the use of WiFi and BT together.  Do you want me to do that?

We also need to wait for releases/tags of pico-sdk and btstack.

dpgeorge committed:
lib/cyw43-driver: Update driver to latest version v1.0.1.

Includes a fix for combined BT+WiFi when using SPI transport.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
lib/btstack: Update to v1.5.6.2.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
lib/pico-sdk: Update to version 1.5.1.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Thanks @peterharperuk.  I've now cleaned up the commits here (squashing the relevant ones), and have also added commits to update to the latest submodules for pico-sdk, cyw43-driver and btstack.

I tested it and everything works as far as I can tell.  In particular there are no more issues with BT and WiFi running at the same time on Pico W.

Apart from waiting for tags on the three submodules, this is ready to go in.

dpgeorge committed:
tools/ci.sh: Build PICO_W board as part of rp2 CI.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk commented:
Thanks for your help, and all the work!

robert-hh commented:
Thank you for all the hard work. I have a question. When updating the submodules with that PR, commit 825a95713 shows a diff:
```
diff --git a/lib/btstack b/lib/btstack
index 1635e36d0..77e752abd 160000
--- a/lib/btstack
+++ b/lib/btstack
@@ -1 +1 @@
-Subproject commit 1635e36d06821af8b61302509e91bfcc1ade84c4
+Subproject commit 77e752abd6a0992334047a48038a5a3960e5c6bc
```
However,. git submodules lists:
`77e752abd6a0992334047a48038a5a3960e5c6bc lib/btstack (v1.4-2373-g77e752abd)`
Same tag, but different version number.

ihew commented:
Guys, while just sat here observing progress, thanks for all the hard and pro work. Really approciated!

Well deserved official praise on [The official Site](https://www.raspberrypi.com/news/new-functionality-bluetooth-for-pico-w/), BTW

dpgeorge commented:
> Same hash value, but different version number.

You probably just need to run `git pull --tags` in `lib/btstack`.  Then `git describe --tags` will work.

samveen commented:
The latest nightly build (1.20+206) includes this. _Thank you for the Music et. al._

robert-hh commented:
> You probably just need to run git pull --tags in lib/btstack

That's what I did, and git tags returns all tags:
```
robert@hh3:~/Downloads/MicroPython/micropython/lib/btstack$ git tag
v1.1
v1.2
v1.2.1
v1.3
v1.3.2
v1.4
v1.4.1
v1.5.0
v1.5.1
v1.5.2
v1.5.3
v1.5.4
v1.5.5
v1.5.5-extra
v1.5.6
v1.5.6.1
v1.5.6.2
v1.5.6.2-extra
robert@hh3:~/Downloads/MicroPython/micropython/lib/btstack$ git describe --tags
v1.5.6.2
```
I tried as well a deinit of btstack and a new init. No change.

robert-hh commented:
Thanks. I was just a little bit nervous about the state of my local clone.
